minigui/minigui.ts

```diff
@@ -0,0 +1,436 @@
+  qBoard,
+  nBoard,
+];
+for (let board of boards) {
+  board.draw();
+}
+
+// Initialize socket connection to the backend.
+let gtp = new gtpsock.Socket(
+  `http://${document.domain}:${location.port}/minigui`,
+  (line: string) => { log.log(line, 'log-cmd'); });
+
+gtp.onConnect(() => {
+  gtp.send('clear_board');
+  gtp.send('mg_gamestate');
+  gameState.gameOver = false;
+  util.getElement('ui-container').classList.remove('hidden');
+});
+
+let qGraph = new Graph('q-graph');
+qGraph.onMoveChanged((move: number | null) => {
+  gameState.hoveredMove = move;
+  if (move == null) {
+    move = gameState.history.length - 1;
+  }
+  let h = gameState.history[move];
+  mainBoard.clearMarks();
+  if (h.lastMove != null && h.lastMove != 'pass' && h.lastMove != 'resign') {
+    mainBoard.setMark(h.lastMove, '#ef6c02');
+  }
+  mainBoard.setStones(h.stones);
+  qBoard.setStones(h.stones);
+  nBoard.setStones(h.stones);
+  mainBoard.setVariation(h.principleVariation);
+  qBoard.setHeatMap(h.q);
+  nBoard.setHeatMap(h.n);
+});
+
+enum Controller {
+  Human,
+  Minigo,
+}
+
+class Player {
+  color: board.Color;
+  controller = Controller.Human;
+  thinking = false;
+
+  constructor(color: board.Color, buttonId: string) {
+    this.color = color;
+    let button = util.getElement(buttonId);
+    button.addEventListener('click', () => {
+      if (this.controller == Controller.Human) {
+        this.controller = Controller.Minigo;
+      } else {
+        this.controller = Controller.Human;
+      }
+      button.innerHTML = `${Controller[this.controller]}`;
+      gtp.send('mg_gamestate');
+    });
+  }
+
+  think() {
+    if (this.controller != Controller.Minigo) {
+      throw new Error(`Unexpected controller: ${Controller[this.controller]}`);
+    }
+    this.thinking = true;
+    gtp.send('mg_genmove 20', (cmd: string, result: string) => {
+      this.thinking = false;
+      onMovePlayed(result);
+    });
+  }
+}
+
+// Callback invoked when either a human or Minigo plays a valid move.
+function onMovePlayed(move: string) {
+  let gameOver = false;
+  if (move == 'pass') {
+    if (++gameState.numConsecutivePasses == 2) {
+      gameOver = true;
+    }
+  } else {
+    gameState.numConsecutivePasses = 0;
+    if (move == 'resign') {
+      gameOver = true;
+    }
+  }
+
+  log.scroll();
+  gtp.send('mg_gamestate');
+
+  if (gameOver && !gameState.gameOver) {
+    gameState.gameOver = true;
+
+    gtp.send('final_score', (cmd: string, result: string) => {
+      let prettyResult: string;
+      if (result[0] == 'W') {
+        prettyResult = 'White wins by ';
+      } else {
+        prettyResult = 'Black wins by ';
+      }
+      if (result[2] == 'R') {
+        prettyResult += 'resignation';
+      } else {
+        prettyResult += result.substr(2) + ' points';
+      }
+
+      let img = document.createElement('img');
+      img.onload = () => { log.scroll(); };
+      img.src = `img/${1 + Math.floor(7 * Math.random())}.jpg`;
```

Comment:
somewhere you should say what you expect the images to be named without making someone track this down.